Persist states

paulyoong opened the pull request.

Commit:
feat(persist child states): use etcd as the store

When a child changes state, persist the new state to the store. This
ensures that when a nexus is recreated it can check the state of the
children before deciding whether or not to use them.

Currently only healthy children are added to the nexus on creation. The
expectation is that the control plane will eventually replace the
unhealthy children in order to satisfy the desired replica count.

Commit:
chore(tidy up): persisting child states

Provide a function which can be used to persist child states to the
store. For now this function continually tries to store the child states
on failure.

Add additional persistent store errors which are used when there is a
failure to wait on a channel for a completed store operation.

Commit:
feat(store endpoint arg): new mayastor argument

Add support for passing the etcd endpoint to mayastor as an argument.

Commit:
refactor(healthy children): check new children

Enhance the "get_healthy_children" function to check for the scenario
where only some of the children can be found in the store. In this case,
only use the healthy children that were found in the store.

Commit:
feat(moac): liveness and readiness probes

Resolves: CAS-548

Commit:
feat(moac): get the default namespace from pod's filesystem

Resolves: CAS-901

tiagolobocastro committed:
chore: add etcd depoyment via helm

Added a helm dependency to our mayastor chart.
Updated the generate yaml script to include the generated etcd yaml files.
Added the newly generated develop yaml files to the deploy folder.

Commit:
test(persistence): add test for persistence

The test checks that the child state is stored to etcd when it is
retired. It is marked as "ignored" for now because it is waiting on
other changes before it will succeed.

Added an additional proto file. This is a draft version of the structure
that will be store in etcd. It is defined in a proto file so that both
Mayastor and MOAC can use it.

Commit:
chore(add files): add missing files

Commit:
refactor(store key): use child uuid

Extract the child UUID from its name and use this as the key for the
child state in the persistent store.

Commit:
refactor(tokio runtime): spawn tasks on tokio

Change the way the persistent store spawns tasks onto a tokio runtime.
Use the new way of spawning futures onto tokio. This greatly simplifies
the code.

The persistent store itself schedules the various operation onto a tokio
runtime. So the store operation can be directly called from Mayastor.

Commit:
refactor(store method): return etcd

The store method has been modified so that it returns an instance of the
etcd client.

Commit:
feat(persistent struct): define persistent info

The data structure stored in the persistent store has been defined in a
new persistence.proto file. The reason for defining it here is because
this structure will be common to MOAC and Mayastor.

A clean shutdown bit is also included in this structure. When a nexus is
destroyed, the clean shutdown variable is set to true. In the event of a
crash, this variable will be false.

On nexus creation, the data structure in the persistent store is
overwritten with the latest information.

Commit:
refactor(nexus uuid): drop "nexus-" prefix

The persistent store entries should have a key which is the nexus UUID.
Drop the "nexus-" prefix from the name before using it as a key.

Changed the PersistOp "Init" enum to "Create".

Commit:
test(new tests): additional persistence tests

Add more tests checking for correct persisted information.

Fixed a bug when updating the state of a child in the persistent store.
Only the child that is changing state is updated in the in-memory
structure. This prevents adding duplicate entries to the vector.

Commit:
fix(moac): moac should handle node removals better

If msn is removed from the cluster (can happen when mayastor on the
node restarts), then certain operations in volume and volume operator
can fail trying to reference the node structure that is being
disassociated with the object (nexus/replica). So handle these cases
gracefully.

Resolves: CAS-902

Commit:
refactor(tidy up): clean up test cases

Simplify the test cases.

Add an additional "mayastor_sleep" function which allows an async sleep
to be called from within Mayastor.

Bug fix: If the etcd endpoint is not provided, do not initialise the
PERSISTENT_STORE once cell. This will default it to None.

chriswldenyer committed:
test(e2e): re-enable XRay

now that XRay failures do not fail the build.

Commit:
refactor(lock): use async lock

Use an async lock to protect the nexus info.

Commit:
Merge #890

890: test(e2e): re-enable XRay r=chriswldenyer a=chriswldenyer

now that XRay failures do not fail the build.

Co-authored-by: chriswldenyer <christopher.denyer@mayadata.io>

Commit:
feat(moac): update dependencies etc

The dependency update fixes a couple of sec vulnerabilities.

We switched back to official k8s client as all our fixes have been
merged to the upstream.

dirty-chai was removed because it's not possible to write TS tests
(definitions are different than expected). Instead we ignore the
eslint error which was why we started to use dirty-chai originally.

Commit:
chore(tidy up): clean up code

tiagolobocastro commented on mayastor/src/bdev/nexus/nexus_persistence.rs:
```diff
@@ -0,0 +1,140 @@
+                }
+                Err(e) => {
+                    tracing::error!(
+                        "Failed to persist info {:?} with error {}. Retrying...",
```
Should we supress subsequent messages and just quietly retry?

tiagolobocastro commented on mayastor/src/bin/mayastor.rs:
```diff
@@ -37,6 +38,8 @@ fn start_tokio_runtime(args: &MayastorCliArgs) {
                 grpc::MayastorGrpcServer::run(grpc_address, rpc_address)
                     .boxed(),
             );
+            futures
```
You can just the init here, just like with the mbus above

tiagolobocastro commented on mayastor/src/core/env.rs:
```diff
@@ -121,6 +122,9 @@ pub struct MayastorCliArgs {
     /// List of cores to run on instead of using the core mask. When specified
     /// it supersedes the core mask (-m) argument.
     pub core_list: Option<String>,
+    #[structopt(short = "-p")]
```
"p" should be enough

tiagolobocastro commented on mayastor/src/core/env.rs:
```diff
@@ -717,6 +726,9 @@ impl MayastorEnvironment {
                 )));
             }
             futures.push(Box::pin(subsys::Registration::run()));
+            futures.push(Box::pin(PersistentStore::init(
```
same as before, can just connect here

tiagolobocastro commented on mayastor/src/persistent_store.rs:
```diff
@@ -0,0 +1,182 @@
+    /// Connect to etcd as the backing store.
+    /// This must be called from a tokio thread context.
+    async fn connect_to_backing_store(endpoint: &str) -> Option<Etcd> {
+        let mut retries = 3;
```
the nats connection retries forever, might want to do the same thing for consistency, not sure...

tiagolobocastro commented on mayastor/src/persistent_store.rs:
```diff
@@ -0,0 +1,182 @@
+            .context(GetWait {
+                key: key.to_string(),
+            })?
+            .map(|r| r.unwrap())
```
what is r?

tiagolobocastro reviewed (commented):
:+1:

Commit:
Merge #885

885: chore: add etcd depoyment via helm r=tiagolobocastro a=tiagolobocastro

Added a helm dependency to our mayastor chart.
Updated the generate yaml script to include the generated etcd yaml files.
Added the newly generated develop yaml files to the deploy folder.

Co-authored-by: Tiago Castro <tiagolobocastro@gmail.com>

Commit:
fix(nexus): failed io may not complete on the same core

Commit:
Merge #889

889: feat(moac): get the default namespace from pod's filesystem r=jkryl a=jkryl

Resolves: CAS-901

Co-authored-by: Jan Kryl <jan.kryl@mayadata.io>

Commit:
chore(review changes): address review comments

Commit:
Merge #891

891: feat(moac): liveness and readiness probes r=jkryl a=jkryl

Resolves: CAS-548

Co-authored-by: Jan Kryl <jan.kryl@mayadata.io>

Commit:
refactor(persisting create): change persist order

Ensure that persisting the create state occurs before setting the nexus
to open. Also filter out any nexuses that are in the init state when
responding to a list nexus gRPC call.

Commit:
Merge #892

892: fix(moac): moac should handle node removals better r=jkryl a=jkryl

If msn is removed from the cluster (can happen when mayastor on the
node restarts), then certain operations in volume and volume operator
can fail trying to reference the node structure that is being
disassociated with the object (nexus/replica). So handle these cases
gracefully.

Resolves: CAS-902

Co-authored-by: Jan Kryl <jan.kryl@mayadata.io>

Commit:
chore(tidy up): error messages

Tidy up the 'put' error messages so that the values are now printed as
strings which makes them more intelligible.

Added info messages for each store operation so that we can easily see
when operations succeed.

Commit:
fix(panic): replace expect

In the trampoline code, if the receiver channel is dropped output an
error message recording the fact rather than panicking.

Commit:
Merge #895

895: feat(moac): update dependencies etc r=jkryl a=jkryl

The dependency update fixes a couple of sec vulnerabilities.

We switched back to official k8s client as all our fixes have been
merged to the upstream.

dirty-chai was removed because it's not possible to write TS tests
(definitions are different than expected). Instead we ignore the
eslint error which was why we started to use dirty-chai originally.

Co-authored-by: Jan Kryl <jan.kryl@mayadata.io>

Commit:
feat(etcd reconnect): retry connections

When etcd is down, the etcd-client library will hold on to any
outstanding requests and waits for a TCP timeout of 15 minutes before
re-establishing a connection. Rather than waiting for this, add a 10
second timeout to each store operation and try to create a new
connection to etcd if it times out.

Commit:
refactor(tidy up): persistent store

Tidy up the persistent store code.

Commit:
test(compose): additional test

Add a test for when the connection to etcd is faulty.

Commit:
chore: remove commented out code

Commit:
refactor(timeout): increase store op timeout

Increase the timeout time for store operations. Also make this time a
constant definition.
